File: app/services/platforms/lemmy.py
import json

import httpx

from app.models.platform_account import PlatformAccount
from app.models.post import Post
from app.services.crypto import decrypt, encrypt
from app.services.platforms.base import BasePlatformClient


class LemmyClient(BasePlatformClient):
    async def post(self, account: PlatformAccount, post: Post) -> str:
        instance = account.instance_url or "https://lemmy.world"
        token = await self._get_token(account, instance)

        # Extra data may contain community_id
        extra: dict = {}
        if account.extra_data_encrypted:
            extra = json.loads(decrypt(account.extra_data_encrypted))

        community_id = extra.get("community_id", 2)  # default to main community

        payload = {
            "name": post.video_title or post.content[:200],
            "body": post.content,
            "community_id": community_id,
            "nsfw": False,
            "auth": token,
        }

        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{instance}/api/v3/post",
                json=payload,
            )
            resp.raise_for_status()
            data = resp.json()
            return str(data["post_view"]["post"]["id"])

    async def get_metrics(self, account: PlatformAccount, platform_post_id: str) -> dict:
        instance = account.instance_url or "https://lemmy.world"

        async with httpx.AsyncClient() as client:
            resp = await client.get(
                f"{instance}/api/v3/post",
                params={"id": platform_post_id},
            )
            resp.raise_for_status()
            data = resp.json()

        counts = data.get("post_view", {}).get("counts", {})
        return {
            "likes": counts.get("score", 0),
            "reposts": 0,
            "replies": counts.get("comments", 0),
            "views": counts.get("hot_rank", 0),
            "clicks": 0,
            "raw_data": json.dumps(counts),
        }

    async def _get_token(self, account: PlatformAccount, instance: str) -> str:
        # Lemmy uses short-lived JWTs; re-authenticate using stored credentials
        extra: dict = {}
        if account.extra_data_encrypted:
            extra = json.loads(decrypt(account.extra_data_encrypted))

        username = extra.get("username")
        password = extra.get("password")
        if not username or not password:
            return decrypt(account.access_token_encrypted)

        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{instance}/api/v3/user/login",
                json={"username_or_email": username, "password": password},
            )
            resp.raise_for_status()
            data = resp.json()
            jwt = data["jwt"]

        # Update stored token
        account.access_token_encrypted = encrypt(jwt)
        return jwt

    @classmethod
    async def connect(cls, instance_url: str, username: str, password: str) -> tuple[str, str, dict]:
        """Returns (jwt, account_name, extra_data)."""
        async with httpx.AsyncClient() as client:
            resp = await client.post(
                f"{instance_url}/api/v3/user/login",
                json={"username_or_email": username, "password": password},
            )
            resp.raise_for_status()
            data = resp.json()
            jwt = data["jwt"]

        # Get user info
        async with httpx.AsyncClient() as client:
            me_resp = await client.get(
                f"{instance_url}/api/v3/site",
                headers={"Authorization": f"Bearer {jwt}"},
            )
            person = me_resp.json().get("my_user", {}).get("local_user_view", {}).get("person", {})
            account_name = person.get("name", username)
            account_id = str(person.get("id", ""))

        extra = {"username": username, "password": password, "account_id": account_id}
        return jwt, account_name, extra

File: src/services/platforms/lemmy.ts
import type { PlatformClient, PostPayload, PostResult, Metrics } from "./types";

interface LemmyCredentials {
  username: string;
  password: string;
  communityId: number;
}

export class LemmyClient implements PlatformClient {
  constructor(
    private readonly instanceUrl: string,
    private readonly credentials: LemmyCredentials,
  ) {}

  private async getToken(): Promise<string> {
    const res = await fetch(`${this.instanceUrl}/api/v3/user/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ username_or_email: this.credentials.username, password: this.credentials.password }),
    });
    if (!res.ok) throw new Error(`Lemmy login failed: ${res.status}`);
    const data = await res.json() as { jwt: string };
    return data.jwt;
  }

  async post(payload: PostPayload): Promise<PostResult> {
    const jwt = await this.getToken();
    const lines = payload.content.split("\n");
    const name = lines[0].slice(0, 200);
    const body = lines.slice(2).join("\n") || payload.content;
    const res = await fetch(`${this.instanceUrl}/api/v3/post`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${jwt}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ name, body, community_id: this.credentials.communityId }),
    });
    if (!res.ok) throw new Error(`Lemmy post failed: ${res.status}`);
    const data = await res.json() as { post_view: { post: { id: number } } };
    return { platformPostId: String(data.post_view.post.id) };
  }

  async getMetrics(platformPostId: string): Promise<Metrics> {
    const res = await fetch(`${this.instanceUrl}/api/v3/post?id=${platformPostId}`);
    if (!res.ok) return { likes: 0, shares: 0, comments: 0, views: 0, clicks: 0, rawData: {} };
    const data = await res.json() as { post_view: { counts: { score: number; comments: number } } };
    const counts = data.post_view.counts;
    return {
      likes: counts.score,
      shares: 0,
      comments: counts.comments,
      views: 0,
      clicks: 0,
      rawData: data as unknown as Record<string, unknown>,
    };
  }
}
